AdblockPlusSafari/AdblockPlusExtras.m

```diff
@@ -0,0 +1,326 @@
+    self.needsDisplayErrorDialog = anyLastUpdateFailed;
+    [self displayErrorDialogIfNeeded];
+  }
+}
+
+- (void)setNeedsDisplayErrorDialog:(BOOL)needsDisplayErrorDialog
+{
+  _needsDisplayErrorDialog = needsDisplayErrorDialog;
+  [self.adblockPlusDetails setBool:needsDisplayErrorDialog forKey:AdblockPlusNeedsDisplayErrorDialog];
+  [self.adblockPlusDetails synchronize];
+}
+
+- (BOOL)anyLastUpdateFailed
+{
+  return [[[self.filterLists allValues] valueForKeyPath:@"@sum.lastUpdateFailed"] integerValue] > 0;
+}
+
+#pragma mark -
+
+- (void)setEnabled:(BOOL)enabled reload:(BOOL)reload
+{
+  self.enabled = enabled;
+
+  if (reload) {
+    [self reloadContentBlockerWithCompletion:nil];
+  }
+}
+
+- (void)setAcceptableAdsEnabled:(BOOL)enabled reload:(BOOL)reload
+{
+  self.acceptableAdsEnabled = enabled;
+
+  if (reload) {
+    [self reloadContentBlockerWithCompletion:nil];
+  }
+}
+
+- (void)reloadContentBlockerWithCompletion:(void(^__nullable)(NSError * __nullable error))completion;
+{
+  __weak __typeof(self) wSelf = self;
+  wSelf.reloading = YES;
+  [SFContentBlockerManager reloadContentBlockerWithIdentifier:self.contentBlockerIdentifier completionHandler:^(NSError *error) {
+    NSLog(@"%@", error);
+    dispatch_async(dispatch_get_main_queue(), ^{
+      wSelf.reloading = NO;
+      [wSelf checkActivatedFlag];
+      if (completion) {
+        completion(error);
+      }
+    });
+  }];
+}
+
+- (void)checkActivatedFlag
+{
+  BOOL activated = [self.adblockPlusDetails boolForKey:AdblockPlusActivated];
+  if (self.activated != activated) {
+    self.activated = activated;
+  }
+}
+
+- (void)updateFilterlists:(BOOL)userTriggered
+{
+  NSMutableDictionary *filterLists = [self.filterLists mutableCopy];
+  for (NSString *filterListName in self.filterLists) {
+    NSURL *url = [NSURL URLWithString:filterListName];
+
+    NSURLSessionTask *task = [self.backgroundSession downloadTaskWithURL:url];
+
+    NSMutableDictionary *filterList = [filterLists[filterListName] mutableCopy];
+    filterList[@"updating"] = @YES;
+    filterList[@"taskIdentifier"] = @(task.taskIdentifier);
+    filterList[@"userTriggered"] = @(userTriggered);
+    filterList[@"lastUpdateFailed"] = @NO;
+    filterLists[filterListName] = filterList;
+
+    [self.downloadTasks[filterListName] cancel];
+    // Store key to task cache
+    self.downloadTasks[filterListName] = task;
+
+    [task resume];
+  }
+  self.filterLists = filterLists;
+}
+
+- (void)displayErrorDialogIfNeeded
+{
+  if (!self.needsDisplayErrorDialog) {
+    return;
+  }
+
+  if (UIApplication.sharedApplication.applicationState != UIApplicationStateActive) {
+    return;
+  }
+
+  UIViewController *viewController = UIApplication.sharedApplication.delegate.window.rootViewController;
+  if ([viewController isKindOfClass:[RootController class]]) {
+    RootController *rootController = (RootController *)viewController;
+    if (![rootController.viewControllers.firstObject isKindOfClass:[HomeController class]]) {
```

Comment:
HomeController need to be loaded in hierarchy (message will not display in welcome guide).